nrf-hal-common/src/uarte.rs

```diff
@@ -351,6 +207,168 @@ where
+    uarte.events_endtx.reset();
+    uarte.events_txstopped.reset();
+
+    // Set up the DMA write
+    uarte.txd.ptr.write(|w|
+        // We're giving the register a pointer to the stack. Since we're
+        // waiting for the UARTE transaction to end before this stack pointer
+        // becomes invalid, there's nothing wrong here.
+        //
+        // The PTR field is a full 32 bits wide and accepts the full range
+        // of values.
+        unsafe { w.ptr().bits(tx_buffer.as_ptr() as u32) });
+    uarte.txd.maxcnt.write(|w|
+        // We're giving it the length of the buffer, so no danger of
+        // accessing invalid memory. We have verified that the length of the
+        // buffer fits in an `u8`, so the cast to `u8` is also fine.
+        //
+        // The MAXCNT field is 8 bits wide and accepts the full range of
+        // values.
+        unsafe { w.maxcnt().bits(tx_buffer.len() as _) });
+
+    // Start UARTE Transmit transaction
+    uarte.tasks_starttx.write(|w|
+        // `1` is a valid value to write to task registers.
+        unsafe { w.bits(1) });
+
+    Ok(())
+}
+
+pub(crate) fn uarte_setup<T: Instance>(uarte: &T, mut pins: Pins, parity: Parity, baudrate: Baudrate) {
+    // Select pins
+    uarte.psel.rxd.write(|w| {
+        let w = unsafe { w.pin().bits(pins.rxd.pin()) };
+        #[cfg(feature = "52840")]
+        let w = w.port().bit(pins.rxd.port().bit());
+        w.connect().connected()
+    });
+    pins.txd.set_high().unwrap();
+    uarte.psel.txd.write(|w| {
+        let w = unsafe { w.pin().bits(pins.txd.pin()) };
+        #[cfg(feature = "52840")]
+        let w = w.port().bit(pins.txd.port().bit());
+        w.connect().connected()
+    });
+
+    // Optional pins
+    uarte.psel.cts.write(|w| {
+        if let Some(ref pin) = pins.cts {
+            let w = unsafe { w.pin().bits(pin.pin()) };
+            #[cfg(feature = "52840")]
+            let w = w.port().bit(pin.port().bit());
+            w.connect().connected()
+        } else {
+            w.connect().disconnected()
+        }
+    });
+
+    uarte.psel.rts.write(|w| {
+        if let Some(ref pin) = pins.rts {
+            let w = unsafe { w.pin().bits(pin.pin()) };
+            #[cfg(feature = "52840")]
+            let w = w.port().bit(pin.port().bit());
+            w.connect().connected()
+        } else {
+            w.connect().disconnected()
+        }
+    });
+
+    // Enable UARTE instance
+    uarte.enable.write(|w| w.enable().enabled());
+
+    // Configure
+    let hardware_flow_control = pins.rts.is_some() && pins.cts.is_some();
+    uarte
+        .config
+        .write(|w| w.hwfc().bit(hardware_flow_control).parity().variant(parity));
+
+    // Configure frequency
+    uarte.baudrate.write(|w| w.baudrate().variant(baudrate));
+}
+
+/// Stop an unfinished UART read transaction and flush FIFO to DMA buffer
+pub(crate) fn uarte_cancel_read<T: Instance>(uarte: &T) {
+    uarte.events_rxto.write(|w| w);
```

Comment:
TODO: I added this line, but I don't know why. I need to re-read the datasheet and make sure I wasn't just debugging something odd.